Created ToggleSwitch and TextField reusable components for use in form

MaxOvenden opened the pull request:
The toggle switches and the text input fields have been moved into reusable components to help with scalability as the project progresses

jrhenderson1988 commented on stockflux-securities-master/src/components/reusableComponents/ToggleSwitch.js:
```diff
@@ -0,0 +1,20 @@
+import React from "react";
+import "./ToggleSwitch.css";
+
+const TextField = ({ value, id, checked, onChange }) => {
```
Shouldn't this be `ToggleSwitch`?

MaxOvenden replied:
Oh dear, such a bad mistake. Yes it should...

MaxOvenden replied:
changed now

jrhenderson1988 commented on stockflux-securities-master/src/components/reusableComponents/ToggleSwitch.js:
```diff
@@ -0,0 +1,20 @@
+        className="input-form-toggle"
+        onChange={onChange}
+      />
+      <label className="toggle-switch" htmlFor={id} />
```
This label is empty, should it have a value or is it not meant to be here?

MaxOvenden replied:
This label is empty because it's simply there for CSS styling purposes. It does it this way in a tutorial I found on how to make a toggle switch.

jrhenderson1988 replied:
Just spoke to you about this and had a quick demo so this is fine

jrhenderson1988 commented on stockflux-securities-master/src/components/reusableComponents/ToggleSwitch.js:
```diff
@@ -0,0 +1,20 @@
+
+const TextField = ({ value, id, checked, onChange }) => {
+  return (
+    <div>
```
If the label is removed then you will have a single component (`<input />`) to return so you won't need a `div`. As a tip, if you want to return elements in your components, it's possible to wrap them with `<React.Fragment>...</React.Fragment>`, or better use its shorthand syntax `<>...</>`

MaxOvenden replied:
Okay I've changed the <div>s to be <> fragment tags.

jrhenderson1988 commented on stockflux-securities-master/src/components/reusableComponents/TextField.js:
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import "./TextField.css";
+
+const TextField = ({ value, id, onChange, readOnly }) => {
```
It can be worth using the spread operator (`...`) to grab any other props that may be passed through and use the spread operator again to pass them as props to your `<input>`. It might also be worth adding support for passing through a custom `className` when your component is used, which would be appended to your existing `className`

MaxOvenden replied:
okay done this now

jrhenderson1988 commented on stockflux-securities-master/src/components/InputForm.js:
```diff
@@ -4,6 +4,8 @@ import { Link } from "react-router-dom";
 import "./InputForm.css";
 import { getSecurity } from "../services/SecuritiesService";
 import ErrorMessage from "./ErrorMessage";
+import TextField from "./reusableComponents/TextField";
```
You can just stick these components in the same place as the others. No need for another folder. The name `reusableComponents` doesn't sound quite right to me

MaxOvenden replied:
Okay, sure. I've moved them.

jrhenderson1988 commented on stockflux-securities-master/src/components/TextField.js:
```diff
@@ -0,0 +1,17 @@
+      className={
+        "input-form-input" +
+        (props.readOnly ? " read-only" : "") +
+        ` ${props.className}`
```
Since you're using backticks here which allows for string interpolation, you can simplify the whole `className`:

    className={`input-form-input ${readOnly && "read-only"} ${className}`}

MaxOvenden replied:
I think that will result in an unwanted space at the end if the props are not defined. I'll do a version of that but fix that issue

MaxOvenden replied:
it looks like it doesn't like me using the `condition && value`  method in the className.  I've asked Matt if he knows why but he said he didn't know. 

` readOnly ? " read-only"  : ""`
does work though.

jrhenderson1988 replied:
> I think that will result in an unwanted space at the end if the props are not defined. I'll do a version of that but fix that issue

It will, yes, but you have that anyway in your current version with ` ${props.className}`. If `className` is not explicitly defined when using your component, a trailing space will be left too. I'd be interested to have a look at why the `condition && value` approach isn't working

jrhenderson1988 commented on stockflux-securities-master/src/components/TextField.js:
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+import "./TextField.css";
+
+const TextField = props => {
```
I think it would be better here to specify the arguments you want to receive in the parameters to the component and use the spread operator to grab the rest of the props:

    const TextField = ({ readOnly, className, ...props }) =>

MaxOvenden replied:
yep added these back in

jrhenderson1988 commented on stockflux-securities-master/src/components/ToggleSwitch.js:
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import "./ToggleSwitch.css";
+
+const ToggleSwitch = props => {
```
As discussed in person and as mentioned for the `TextField` component, I think it would be better to specify the values you want to receive in your arguments and use the spread operator to receive the rest:

    const ToggleSwitch = ({ id, onChange, className, ...props }) => {

This effectively extracts `id`, `className` and `onChange` from the `props` and leaves the rest of them intact in the `props` variable so when you spread `...props` into the component in your JSX, those props won't be included and you can manually add them in with your custom logic

MaxOvenden replied:
okay , understood. Made the change now

jrhenderson1988 commented on stockflux-securities-master/src/components/ToggleSwitch.js:
```diff
@@ -0,0 +1,18 @@
+      <input
+        {...props}
+        type="checkbox"
+        className={"input-form-toggle" + ` ${props.className}`}
```
Similar to in the `TextField` component here, if you expect these properties in the component arguments directly, you can remove the `props.` prefix since they will exist as their own variables. You can also use the backtick, string interpolation feature to simply the `className` expression:

    className={`input-form-toggle ${className}`}

MaxOvenden replied:
Yep I've done this.

Worth noting:  I've moved the extra className from the input to the label as the input is hidden so extra css won'yt do anything to it. The label is the one which is actually displayed

jrhenderson1988 replied:
Cool. Yeah, that's the right approach since the input is hidden anyway.

MaxOvenden commented:
> Just a few minor things. As discussed as well, it would be worth using the React PropTypes library to define the types of the props accepted by each component and whether or not they're mandatory or optional.

I've addressed these changes and added React PropTypes to both components. However, note my comment about not being able to use the && method in the classNames.

jrhenderson1988 commented on stockflux-securities-master/src/components/InputForm.js:
```diff
@@ -102,40 +96,34 @@ const InputForm = ({ match }) => {
               checked={visible}
-              type="checkbox"
               onChange={() => setVisible(!visible)}
+              className="testClass"
```
Did you mean to remove this?

MaxOvenden replied:
woops, yes I did

MaxOvenden replied:
gone now

jrhenderson1988 commented on stockflux-securities-master/src/components/TextField.js:
```diff
@@ -0,0 +1,28 @@
+};
+
+TextField.propTypes = {
+  onChange: PropTypes.func.isRequired,
```
I'm not sure that any of these should be required. I think it's totally valid to be able to omit an `onChange` function and an `id`. A `value` is only really necessary if an existing value is actually present. In most cases they will be provided but I don't think they should be made mandatory. It might actually be worth removing these 3 fields from the destructured arguments to the component and just allowing the `{...props}` spread to apply them on the input element. The prop types you've defined should take care of ensuring that the correct types are given when the component is used.

MaxOvenden replied:
yep done

jrhenderson1988 commented on stockflux-securities-master/src/components/ToggleSwitch.js:
```diff
@@ -0,0 +1,28 @@
+ToggleSwitch.propTypes = {
+  onChange: PropTypes.func.isRequired,
+  id: PropTypes.string.isRequired,
+  value: PropTypes.bool.isRequired,
```
Value shouldn't have to be a boolean type. Checkbox input fields are often given string values.

MaxOvenden replied:
I've made id required as without it the component doesn't function. I've made value and onChange optional though
